tokio-trace-core/src/field.rs

```diff
@@ -30,74 +34,256 @@
+        self.close_list()
+    }
 
-/// Trait implemented by types which may be converted into a `Value`.
-///
-/// Implementors of `AsValue` must provide an implementation of the `fmt_value`
-/// function, which describes how to format a value of this type.
-pub trait AsValue: Send + Sync {
-    /// Formats the value with the given formatter.
-    fn fmt_value(&self, f: &mut fmt::Formatter) -> fmt::Result;
-
-    #[doc(hidden)]
-    fn type_id(&self) -> TypeId
+    /// Record a `struct` list of `Value`s, given the struct's `name` and an
+    /// iterator over its `fields`.
+    ///
+    /// This function manages calling `open_struct`, recording the struct's
+    /// fields, and closing the struct.
+    ///
+    /// This is the suggested way for `Value` implementations to record structs,
+    /// rather than calling those functions directly, unless different behaviour
+    /// is needed.
+    pub fn record_struct<'a, I>(&mut self, name: &str, fields: I) -> RecordResult
     where
-        Self: 'static,
+        I: IntoIterator<Item = (&'a str, &'a dyn Value)>,
     {
-        TypeId::of::<Self>()
+        self.open_struct(name)?;
+        for (name, v) in fields {
+            self.record_kv(&name, v)?;
+        }
+        self.close_struct()
+    }
+
+    /// Record a tuple.
+    ///
+    /// This function manages calling `open_tuple`, recording the tuple's
+    /// fields, and closing the tuple.
+    ///
+    /// This is the suggested way for `Value` implementations to record tuples,
+    /// rather than calling those functions directly, unless different behaviour
+    /// is needed.
+    pub fn record_tuple<'a, I>(&mut self, i: I) -> RecordResult
+    where
+        I: IntoIterator<Item = &'a dyn Value>,
+    {
+        self.open_tuple()?;
+        for v in i {
+            v.record(self)?;
+        }
+        self.close_tuple()
     }
 }
 
-/// Trait representing a type which may be converted into an `OwnedValue`.
+/// Result type returned by recording field values with a `Record`.
+pub type RecordResult = Result<(), Error>;
+
+/// A structured field value of an erased type.
 ///
-/// References to types implementing `IntoValue` may be formatted (as `Value`s),
-/// _or_ may be converted into owned `OwnedValue`s. In addition to being owned,
-/// instances of `OwnedValue` may also be downcast to their original erased type.
-pub trait IntoValue: AsValue {
-    /// Converts this type into an `OwnedValue`.
-    fn into_value(&self) -> OwnedValue;
+/// Implementors of `Value` may call the appropriate typed recording methods on
+/// the `Record` passed to `Record` in order to indicate how their data
+/// should be recorded.
+pub trait Value: fmt::Debug + Send + Sync {
+    /// Records the value with the given `Record`.
+    fn record(&self, &mut dyn Record) -> RecordResult;
+}
+
+/// An error that occurred while recording a field.
+#[derive(Debug)]
+pub struct Error {
+    // TODO: Should this carry a boxed error, or should it just discard the
```

Comment:
Boxed error, IMO. Certain subscribers _will_ care about the error.

Reply:
Yeah, I didn't want to start discarding it without putting more thought into it, but the opaque struct means we can change how much information this carries w/o a breaking change.

Reply:
I think this is a reasonable design decision right now which can be revisited later, if necessary.